core/src/main/scala/shapeless/generic.scala

```diff
@@ -380,82 +381,85 @@ trait CaseClassMacros extends ReprTypes {
-            }
+        def substituteArgs = {
+          val subst = thisType(sym).baseType(baseSym).typeArgs
+          for (param <- sym.typeParams)
+            yield baseArgs(subst.indexWhere(_.typeSymbol == param))
+        }
+
+        // Handle anonymous subclasses.
+        val isNamed = !isAnonOrRefinement(sym)
+        if (isNamed && sym.isStatic) {
+          if (sym.isModuleClass) sym.toType
+          else appliedType(sym.toType, substituteArgs)
+        } else {
+          // Look for a path from the macro call site to the subtype.
+          val owner = sym.owner
+          val owners = ownerChain(if (isNamed) owner else owner.owner)
+          val prePaths = for (pre <- Iterator.iterate(basePre)(prefix).takeWhile(_ != NoPrefix))
+            yield (pre, owners.iterator.dropWhile(pre.baseType(_) == NoType))
+
+          // Find a path from a (sub-)prefix or the enclosing owner.
+          val (pre0, path) = prePaths.find(_._2.nonEmpty).getOrElse {
+            val enclosing = ownerChain(enclosingOwner)
+            val common = owners zip enclosing indexWhere { case (o1, o2) => o1 != o2 }
+            (NoPrefix, if (common < 0) Iterator.empty else owners.iterator drop common - 1)
           }
 
-          val suffix = ownerChain(sym).dropWhile(_ != basePre.typeSymbol)
-          val ctor =
-            if(suffix.isEmpty) {
-              if(sym.isModuleClass) {
-                val moduleSym = sym.asClass.module
-                val modulePre = prefix(moduleSym.typeSignatureIn(basePre))
-                c.internal.singleType(modulePre, moduleSym)
-              } else
-                appliedType(sym.toTypeIn(basePre), substituteArgs)
-            } else {
-              if(sym.isModuleClass) {
-                val path = suffix.tail.map(sym => normalizeTermName(sym.name))
-                val (modulePre, moduleSym) = mkDependentRef(basePre, path)
-                c.internal.singleType(modulePre, moduleSym)
-              } else if(isAnonOrRefinement(sym)) {
-                val path = suffix.tail.init.map(sym => normalizeTermName(sym.name))
-                val (valPre, valSym) = mkDependentRef(basePre, path)
-                c.internal.singleType(valPre, valSym)
-              } else {
-                val path = suffix.tail.init.map(sym => normalizeTermName(sym.name)) :+ suffix.last.name.toTypeName
-                val (subTpePre, subTpeSym) = mkDependentRef(basePre, path)
-                c.internal.typeRef(subTpePre, subTpeSym, substituteArgs)
-              }
-            }
-          if(!isAccessible(ctor))
-            abort(s"$tpe has an inaccessible subtype $ctor")
-          if(ctor <:< baseTpe) Some(ctor) else None
+          // Construct a stable prefix from the path.
+          val pre = path.drop(1).foldLeft(pre0) { (pre1, part) =>
+            if (part.isType) part.asType.toTypeIn(pre1)
+            else abort(s"$tpe has a subtype $sym with unstable prefix")
+          }
```

Comment:
At this point we have a (possibly empty) path from the macro call site to the subclass and we can make a prefix out of it. Since we already have all the symbols (in the owner chain), there is no need to fall back to calling `.member` with names like the old implementation. Instead we just make sure all symbols in the path are types (otherwise the subclass is hidden in a val, var or def) and call `toTypeIn`.